src/i18n.cpp

```diff
@@ -0,0 +1,145 @@
+#include "i18n.h"
+#include <QTranslator>
+#include <QLibraryInfo>
+#include <QApplication>
+#include <QSettings>
+#include <QDebug>
+
+namespace {
+const char* langs[] = {
+    NULL, //reserved for system locale
+    "ca",
+    "de_DE",
+    "en",
+    "es",
+    "es_AR",
+    "es_MX",
+    "fr_FR",
+    "he_IL",
+    "hu_HU",
+    "is",
+    "it",
+    "ko_KR",
+    "nl_BE",
+    "pl_PL",
+    "pt_BR",
+    "pt_PT",
+    "ru",
+    "sk_SK",
+    "uk",
+    "zh_CN",
+    "zh_TW",
+    "tr",
+    NULL
+    };
+void saveCurrentLanguage(int langIndex) {
+    QSettings settings;
+
+    settings.beginGroup("Language");
+    settings.setValue("current", QString(langs[langIndex]));
+    settings.endGroup();
+}
+
+int loadCurrentLanguage() {
+    QSettings settings;
+
+    settings.beginGroup("Language");
+    QString current = settings.value("current").toString();
+    settings.endGroup();
+
+    // system locale
+    if (current.isEmpty()) {
+        return 0;
+    }
+
+    const char** pos = langs; /* skip first one*/
+    while(*++pos != NULL) {
+        if (*pos == current)
+            break;
+    }
+    return pos - langs;
+}
+
+} // anonymous namespace
+
+I18NHelper *I18NHelper::instance_ = NULL;
+
+I18NHelper::I18NHelper()
+    : qt_translator_(new QTranslator),
+      my_translator_(new QTranslator)
+{
+    qApp->installTranslator(qt_translator_.data());
+    qApp->installTranslator(my_translator_.data());
+}
+
+I18NHelper::~I18NHelper() {
+}
+
+void I18NHelper::init() {
+    int pos = preferredLanguage();
+    if (langs[pos] == NULL) // NULL is reserved for system locale
+        setLanguage(0);
+    else
+        setLanguage(pos);
+}
+
+int I18NHelper::preferredLanguage() {
+    return loadCurrentLanguage();
+}
+
+void I18NHelper::setPreferredLanguage(int langIndex) {
+    const QList<QLocale> &locales = getInstalledLocales();
+    if (langIndex < 0 || langIndex >= locales.size())
+        return;
+    saveCurrentLanguage(langIndex);
```

Comment:
No, it is saved as a string in QSettings.And the int number is only used at runtime and calculated dynamically from the string.

see more at [saveCurrentLanguage](https://github.com/haiwen/seafile-client/pull/404/files#diff-f0f193787a82c9b195cd700c580219aeR35)